src/OrchardCore.Modules/OrchardCore.Media/Handlers/MediaFieldIndexHandler.cs

```diff
@@ -0,0 +1,62 @@
+using System.Linq;
+using System.Threading.Tasks;
+using OrchardCore.Indexing;
+using OrchardCore.Media.Fields;
+using OrchardCore.Media.Settings;
+using UglyToad.PdfPig;
+
+namespace OrchardCore.Media.Handlers
+{
+    public class MediaFieldIndexHandler : ContentFieldIndexHandler<MediaField>
+    {
+        private readonly IMediaFileStore _mediaFileStore;
+
+        public MediaFieldIndexHandler(IMediaFileStore mediaFileStore)
+        {
+            _mediaFileStore = mediaFileStore;
+        }
+
+        public async override Task BuildIndexAsync(MediaField field, BuildFieldIndexContext context)
+        {
+            var options = context.Settings.ToOptions();
+            var settings = context.ContentPartFieldDefinition.GetSettings<MediaFieldSettings>();
+
+            if (field.Paths.Length > 0)
+            {
+                if (settings.AllowMediaText)
+                {
+                    foreach (var key in context.Keys)
+                    {
+                        foreach (var mediaText in field.MediaTexts)
+                        {
+                            context.DocumentIndex.Set(key, mediaText, options);
+                        }
+                    }
+                }
+
+                foreach (var path in field.Paths.Where(path => path.EndsWith(".pdf")))
```

Comment:
Sebastien's idea: Make this extensible, just via options (not DI).

Reply:
I just recognized I don't know what you meant by options @sebastienros. If `IOptions<T>` then I'm not sure how providing an extension point for services that would hook into this indexing work. I'd rather create an injectable `IMediaFieldFileIndexHandler` and run through them here.

Reply:
@Piedone like the feature, it is the thing that I always recomend to clients not do do, is use too many pdfs, because indexing and seo is dificult with them.

re options: using `Options<T>` vs providers is essentially the same thing, just potentially easier to configure, and less expensive, in terms of DI lookups etc. There are probably more places we could have used the Options pattern, over a DI pattern.

Consider for example this could just be a dictionary of funcs on an options class.

```
public Dictionary<string, Func<Stream, string[]>> MediaPaths;
```

then in the `MediaFieldIndexHandler`

```
if (_options.TryGetValue(fileExtension, out var indexer)
{
   var values = indexer(fileStrem);
}
```

making it trivial to `Configure` a `.txt` file indexer etc al.

Of course exactly the same thing could be done with a provider, so as you prefer.

In terms of the `FullText` index, there are two ways to go here, as you have done, by adding the new fields to the Search Index, or using the `Custom Full Text Settings` liquid editor on a content type, to return these values into the `FullTextIndex`. I tend towards doing that more, as it makes it easier to differentiate between indexes that we might use internally for custom lucene queries, and not have to add so many fields to the index. For example with the `full_text` liquid aspect, we are also able to index into `Widgets` and `BagPart` items, and control which fields on these contained items are used for the full text index, from their own content definitions.

For example media text can already be added to a full text index, simply by returning (in a loop) `{{ Content.MyPart.MyField.MediaTexts }}`

What we also need at some point, is the same feature as the Full Text Index Settings for the `BodyAspect` to allow Widgets and BagParts to better contribute to a `BodyAspect` for rss feeds etc.

Whatever you decide in terms of providers vs options, it would be good, if it was reusable (later or now) in a liquid filter, to achieve the same thing as we can with `| full_text` 

and what @Skrypt says about about the key names

Reply:
Ah, thanks for the clarification on options. I'd prefer the provider-based approach though: While if the configured file content extractor does nothing apart from processing the stream having such delegates in an options class is easier, if it needs to access injected services too it'd become cumbersome for the consumer (or we'd need to pass and `IServiceProvider` around as `IBackgroundTask` does).

As for how to handle FullText:

While handling `MediaText` can be done as follows too:

```liquid
{% for mediaText in Model.Content.ContentTypeName.FieldName.MediaTexts %}
  {{ mediaText }}
{% endfor %}
```

I think doing something similar for file contents would complicate things quite a lot (both on the implementation and configuration side), and be an exception to what fields usually do (which is saving fields in a `ContentFieldIndexHandler<T>`.

In the wider scope I'd probably prefer all fields that are configured to be indexed (and stored and analyzed) to be automatically added to `Content.ContentItem.FullText` as well, so no further configuration is necessary to make it searchable too. So maybe a `FullTextAspect`-populating handler for each content field?

Reply:
I see your point and it kind of makes sense but not at all times. So, I guess we should have some global recipes for Content Types FullText indexing. This would be a global full-text configuration that would be overriden by the `Content Type` settings themselves. I'm not sure if it's a good idea to put everything in the full-text column while you can already search in every field with a Lucene Query. I kind of understand that since every content types have different fields it makes it harder to keep all the Content Types full-text settings to be properly configured when people are adding new fields which should be added to the full-text index. Though, it is wrong to assume that every field should be part of it. So this would require some more thinking.

Reply:
Perhaps a field setting to add it to `FullText` optionally? Thus have a simple way to use them for search but keep the option open for more involved usage in custom queries.